quixstreams/sources/base.py

```diff
@@ -0,0 +1,285 @@
+
+
+logger = logging.getLogger(__name__)
+
+__all__ = (
+    "BaseSource",
+    "Source",
+    "PollingSource",
+)
+
+
+class BaseSource(ABC):
+    """
+    This is the base class for all sources.
+
+    Sources are executed in a sub-process of the main application.
+
+    To create your own source you need to implement:
+        * `run`
+        * `cleanup`
+        * `stop`
+        * `default_topic`
+    """
+
+    # time in seconds the application will wait for the source to stop.
+    shutdown_timeout: float = 10
+
+    def __init__(self):
+        self._producer: Optional[RowProducer] = None
+        self._producer_topic: Optional[Topic] = None
+        self._configured: bool = False
+
+    def configure(self, topic: Topic, producer: RowProducer) -> None:
+        """
+        This method is triggered when the source is registered to the Application.
+
+        It's used to configure the source with the kafka producer and the topic it should use
+        """
+        self._producer = producer
+        self._producer_topic = topic
+        self._configured = True
+
+    @property
+    def configured(self):
+        return self._configured
+
+    @property
+    def producer_topic(self):
+        return self._producer_topic
+
+    @abstractmethod
+    def run(self) -> None:
+        """
+        This method is triggered in the subprocess when the source is started.
+
+        The subprocess will run as long as the run method execute. You should use it to fetch data and produce it to kafka.
+        """
+
+    @abstractmethod
+    def cleanup(self, failed: bool) -> None:
+        """
+        This method is triggered once the `run` method completes.
+
+        You can use it to perform any kind of shutting down cleanup. For example
+        flushing the producer.
+        """
+
+    @abstractmethod
+    def stop(self) -> None:
+        """
+        This method is triggered when the application is shutting down.
+
+        The source should make sure it's `run` method completes soon.
+
+        Example Snippet:
+
+        ```python
+        class MySource(BaseSource):
+            def run(self):
+                self._running = True
+                while self._running:
+                    self._producer.produce(
+                        topic=self._producer_topic,
+                        value="foo",
+                    )
+                    time.sleep(1)
+
+            def stop(self):
+                self._running = False
+        ```
+        """
+
+    @abstractmethod
+    def default_topic(self) -> Topic:
+        """
+        This method is triggered when the user hasn't specified a topic for the source.
+
+        The source should return a default topic configuration
+        """
+
+
+class Source(BaseSource):
+    """
+    BaseSource class implementation providing
+
+    Implementation for the abstract method:
+        * `default_topic`
+
+    Helper methods
+        * serialize
+        * produce
+        * flush
```

Comment:
If we expect users to subclass from the `Source` class, we must also add some basic examples here.